IMOD/manpages/joinmodel.man
.na
.nh
.TH joinmodel 1 4.6.34 IMOD
.SH NAME
joinmodel - to join two models together
.SH SYNOPSIS
joinmodel
.SH DESCRIPTION
Joinmodel will combine two models, joining objects in the
SECOND model to the ends of objects in the FIRST model, if the
corresponding points match up sufficiently well.  Branch points are
allowed in either model, and should be preserved even during joining
of objects.  Point "marks" (numbers controlling point symbol
display), which are properties of individual points in the model,
will be preserved, but text labels in the second model will not.
.P
Objects will be joined together only if there is physical
overlap between the end of the object in the FIRST model and the
start of the object in the SECOND model.  If an object in the FIRST
model ends on one section and an object in the SECOND model starts
in the same location on the next section, they will not be joined.
.P
When two objects are joined together, the points in the object in the
SECOND model that overlap will be thrown away, and the rest of the
points in that object will be added to the end of the object in the
FIRST model.  The resulting object will have the same color and
other attributes as the original object in the FIRST model.  All
other objects will be carried over into the joined model without
modification.
.P
Entries to the program:
.P
.nf
Name of SECOND model file (YES, it needs the SECOND one first).
Name of FIRST model file
Name of output model file
.fi
.P
Maximum distance in the X-Y plane, and maximum separation in Z,
   allowed in order for points from the two different models to be
   considered matching points.  Enter / to accept the default values
   in brackets.
.P
Maximum overlap to look for between objects in the two models.
   Enter / to accept the default value in brackets.  This parameter
   controls how long it takes the program to look for matches.
.P
.P
The program considers the end point of each object in the FIRST
model in turn.  It looks for a match between that end point and the
first, second, etc. point of each object in the SECOND model
(up to the number specified by the maximum overlap parameter).  When
it finds a match, it then tests whether each of the points from the
beginning of the object in the SECOND model up to that matching point
also match the corresponding points in the object in the FIRST model.
Only if there is a complete match along the entire stretch of overlap
will the objects be joined.
.SH HISTORY
.nf
Written by David Mastronarde, 1/10/90
.fi
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/flib/model/joinmodel.f
*       * * * * JOINMODEL * * * * *
c       
c       JOINMODEL will combine two models, joining objects in the
c       SECOND model to the ends of objects in the FIRST model, if the
c       corresponding points match up sufficiently well.  Branch points are
c       allowed in either model, and should be preserved even during joining
c       of objects.  Point "marks" (numbers controlling point symbol
c       display), which are properties of individual points in the model,
c       will be preserved, but text labels in the second model will not.
c       
c       Objects will be joined together only if there is physical
c       overlap between the end of the object in the FIRST model and the
c       start of the object in the SECOND model.  If an object in the FIRST
c       model ends on one section and an object in the SECOND model starts
c       in the same location on the next section, they will not be joined.
c       
c       When two objects are joined together, the points in the object in the
c       SECOND model that overlap will be thrown away, and the rest of the
c       points in that object will be added to the end of the object in the
c       FIRST model.  The resulting object will have the same color and
c       other attributes as the original object in the FIRST model.  All
c       other objects will be carried over into the joined model without
c       modification.
c       
c       Entries to the program:
c       
c       Name of SECOND model file (YES, it needs the SECOND one first).
c       Name of FIRST model file
c       Name of output model file
c       
c       Maximum distance in the X-Y plane, and maximum separation in Z,
c       .  allowed in order for points from the two different models to be
c       .  considered matching points.  Enter / to accept the default values
c       .  in brackets.
c       
c       Maximum overlap to look for between objects in the two models.
c       .  Enter / to accept the default value in brackets.  This parameter
c       .  controls how long it takes the program to look for matches.
c       
c       
c       The program considers the end point of each object in the FIRST
c       model in turn.  It looks for a match between that end point and the
c       first, second, etc. point of each object in the SECOND model
c       (up to the number specified by the maximum overlap parameter).  When
c       it finds a match, it then tests whether each of the points from the
c       beginning of the object in the SECOND model up to that matching point
c       also match the corresponding points in the object in the FIRST model.
c       Only if there is a complete match along the entire stretch of overlap
c       will the objects be joined.
c       
c       David Mastronarde  1/10/90
c       
      use fortmodel
c       
c       Variable definitions for storage of second model
c       
      integer*4, allocatable :: object2(:)             !pointer to p_coord array
      integer*2, allocatable :: npt_in_obj2(:)        !# of points in object
      integer*4, allocatable :: ibase_obj2(:)         !base index of object in OBJECT
      integer*2, allocatable :: obj_color2(:,:)       !ON/OFF and color indexes
      real*4, allocatable :: p_coord2(:,:)                !model coordinates of points
      integer*1, allocatable :: pt_label2(:)    !symbol number for point
      logical, allocatable :: matched(:)

      logical exist,readw_or_imod,extended,doesmatch,failed
      character*80 modelfile
      data maxoverlap/10/,distmax/2./,zsepmax/0.1/
c       
c       get second model first
c       
      call setExitPrefix('ERROR: JOINMODEL - ')
10    write(*,'(1x,a,$)')'Name of SECOND input model file: '
      read(*,'(a)')modelfile
15    exist=readw_or_imod(modelfile)
      if(.not.exist)then
        print *,'no good, try again'
        go to 10
      endif
      less_pt = max_pt/5
      allocate(object2(len_object), npt_in_obj2(max_obj_num), ibase_obj2(max_obj_num),
     &    obj_color2(2,max_obj_num), p_coord2(3,less_pt), pt_label2(less_pt),
     &    matched(max_obj_num), stat = ierr)
      call memoryError(ierr, 'arrays for second model')
c       
c       repack this model before storing in second arrays
c       
      call repack_mod
c       
c       store in second arrays
c       
      if (n_point.gt.less_pt) stop 'MODEL TOO LARGE FOR ARRAYS'
      nobj=n_object
      nintot=n_point
      max_obj2=max_mod_obj
c       
      do i=1,n_point
        do j=1,3
          p_coord2(j,i)=p_coord(j,i)
        enddo
        pt_label(i)=pt_label(i)
      enddo
c       
      do iobj=1,max_mod_obj
        npt_in_obj2(iobj)=npt_in_obj(iobj)
        ibase_obj2(iobj)=ibase_obj(iobj)
        obj_color2(1,iobj)=obj_color(1,iobj)
        obj_color2(2,iobj)=obj_color(2,iobj)
      enddo
c       
      do i=1,len_object
        object2(i)=object(i)
      enddo
c       
c       get model being added to
c       
20    write(*,'(1x,a,$)')'Name of FIRST input model file: '
      read(*,'(a)')modelfile
25    exist=readw_or_imod(modelfile)
      if(.not.exist)then
        print *,'no good, try again'
        go to 20
      endif
c       
c       repack this model before working with it
c       
      call repack_mod
c       
      write(*,'(1x,a,$)')'Name of output model file: '
      read(*,'(a)')modelfile
c       
      write(*,'(1x,a,/,a,f3.1,a,f3.1,a,$)')'Maximum distance in X-Y'//
     &    ' plane and maximum separation in Z allowed',
     &    '   between points to consider them matching [',
     &    distmax,',',zsepmax,']: '
      read(*,*)distmax,zsepmax
      distsqr=distmax**2
c       
      write(*,'(1x,a,i2,a,$)')'Maximum overlap to seek'//
     &    ' between ends of objects [',maxoverlap,']: '
      read(*,*)maxoverlap
c       
c       go through objects in first model, comparing last point of each to
c       points in all objects in second model
c       

      do iobj=1,max_mod_obj
        ninobj1=npt_in_obj(iobj)
        ibase=ibase_obj(iobj)
        if(ninobj1.gt.0)then
          jpnt=abs(object(ninobj1+ibase))
          xx=p_coord(1,jpnt)
          yy=p_coord(2,jpnt)
          zz=p_coord(3,jpnt)
          extended=.false.
          job=1
          do while (.not.extended.and.job.le.max_obj2)
            ibase2=ibase_obj2(job)
            ninobj2=npt_in_obj2(job)
            if(ninobj2.gt.0.and..not.matched(job))then
              ip=1
              do while (.not.extended.and.
     &            ip.le.min(maxoverlap,ninobj2))
                ipnt2=abs(object2(ip+ibase2))
                if(abs(p_coord2(3,ipnt2)-zz).le.zsepmax)then
                  dx=abs(xx-p_coord2(1,ipnt2))
                  dy=abs(yy-p_coord2(2,ipnt2))
                  if(dx.le.distmax.and.dy.le.distmax.and.
     &                dx**2+dy**2.le.distsqr)then
c                     
c                     got a match, now look back from 1 to ip-1 for match
c                     
                    if(ninobj1.ge.ip)then
                      doesmatch=.true.
                      ipm=1
                      do while (doesmatch.and.ipm.lt.ip)
                        jpnt=abs(object(ibase+ninobj1+ipm-ip))
                        ipnt2=abs(object2(ipm+ibase2))
                        doesmatch=abs(p_coord2(3,ipnt2)-
     &                      p_coord(3,jpnt)) .le.zsepmax .and.
     &                      (p_coord2(1,ipnt2)-p_coord(1,jpnt))**2+
     &                      (p_coord2(2,ipnt2)-p_coord(2,jpnt))**2
     &                      .le.distsqr
                        ipm=ipm+1
                      enddo
c                       
c                       if everything matches, now add from ip+1 to end
c                       
                      if(doesmatch)then
                        call object_mover(iobj,failed)
                        if(failed)stop 'insufficient object space'
                        if(ninobj2-ip+n_point.ge.max_pt)
     &                      stop 'insufficient point space'
                        ibase=ibase_obj(iobj)
c                         
                        do iadd=ip+1,ninobj2
                          ipnt2=object2(ibase2+iadd)
                          ninobj1=ninobj1+1
                          if(ipnt2.gt.0)then
c                             
c                             if not a branch point, just add point on end
c                             
                            n_point=n_point+1
                            ntot_in_obj=ntot_in_obj+1
                            p_coord(1,n_point)=p_coord2(1,ipnt2)
                            p_coord(2,n_point)=p_coord2(2,ipnt2)
                            p_coord(3,n_point)=p_coord2(3,ipnt2)
                            object(ibase+ninobj1)=n_point
                            pt_label(n_point)=pt_label2(ipnt2)
                          else
c                             
c                             otherwise, find positive reference at prior
c                             place in this object, and use that to set up
c                             branch reference
c                             
                            iref=0
                            look=1
                            do while (look.lt.iadd .and. iref.eq.0)
                              if(object2(ibase2+look).eq.-ipnt2)
     &                            iref=look
                              look=look+1
                            enddo
                            if(iref.gt.0)then
                              object(ibase+ninobj1)=-object(ibase+
     &                            iref-iadd+ninobj1)
                            else
                              print *,'unresolvable branch reference',
     &                            ' in second model, object',job,
     &                            ', point',iadd
                              stop
                            endif
                          endif
                        enddo
c                         
                        npt_in_obj(iobj)=ninobj1
                        ibase_free=max(ibase_free,ibase+ninobj1)
3                       matched(job)=.true.
                        extended=.true.
                      endif
                    endif
                  endif
                endif
                ip=ip+1
              enddo                             !end of loop through points
            endif
            job=job+1
          enddo                                 !end of loop thru 2nd objects
        endif
      enddo
c       
c       now take unmatched objects of 2nd model and make them new objects
c       first call the object packer, because it won't get called ever
c       when new objects are being allocated
c       
      call object_packer
      iobject=1
      do job=1,max_obj2
        if(.not.matched(job).and.npt_in_obj2(job).gt.0)then
c           
c           find first free object
c           
          do while (npt_in_obj(iobject).gt.0.and.
     &        iobject.le.max_obj_num)
            iobject=iobject+1
          enddo
          call object_mover(iobject,failed)
          if(failed)stop 'insufficient object space'
          if(n_point+npt_in_obj2(job).ge.max_pt)
     &        stop 'insufficient point space'
c           
c           take care of all the counters and pointers
c           
          n_object=n_object+1
          if(n_object.gt.max_obj_num)stop 'too many objects'
          ibase_obj(iobject)=ibase_free
          nin_order=nin_order+1
          obj_order(nin_order)=iobject
          ndx_order(iobject)=nin_order
          ibase=ibase_free
          max_mod_obj=max(max_mod_obj,iobject)
c           
c           move the points into model arrays
c           
          ibase2=ibase_obj2(job)
          do iadd=1,npt_in_obj2(job)
            ipnt2=object2(ibase2+iadd)
            if(ipnt2.gt.0)then
c               
c               if not a branch point, just add point on end
c               
              n_point=n_point+1
              ntot_in_obj=ntot_in_obj+1
              p_coord(1,n_point)=p_coord2(1,ipnt2)
              p_coord(2,n_point)=p_coord2(2,ipnt2)
              p_coord(3,n_point)=p_coord2(3,ipnt2)
              pt_label(n_point)=pt_label2(ipnt2)
              object(ibase+iadd)=n_point
            else
c               
c               otherwise, find positive reference at prior place in this
c               object, and use that to set up branch reference
c               
              iref=0
              look=1
              do while (look.lt.iadd .and. iref.eq.0)
                if(object2(ibase2+look).eq.-ipnt2) iref=look
                look=look+1
              enddo
              if(iref.gt.0)then
                object(ibase+iadd)=-object(ibase+iref)
              else
                print *,'unresolvable branch reference',
     &              ' in second model, object',job,', point',iadd
                stop
              endif
            endif
          enddo
          npt_in_obj(iobject)=npt_in_obj2(job)
          obj_color(1,iobject)=obj_color2(1,job)
          obj_color(2,iobject)=obj_color2(2,job)
          ibase_free=ibase+npt_in_obj2(job)
        endif
      enddo
c       
c       now just write model
c       
77    call write_wmod(modelfile)
      print *, 'JOINED MODEL WRITTEN'
      call exit(0)
      end
